state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -383,102 +392,17 @@ pub mod pallet {
-
-				// Swap into Stable asset first.
-				Self::do_group_and_swap(&mut swaps, SwapLeg::ToStable)?;
-
-				// Take NetworkFee for all swaps
-				for swap in swaps.iter_mut() {
-					debug_assert!(
-						swap.stable_amount.is_some(),
-						"All swaps should have Stable amount set here"
-					);
-					let stable_amount = swap.stable_amount.get_or_insert_with(Default::default);
-					*stable_amount = T::SwappingApi::take_network_fee(*stable_amount);
-				}
 
-				// Swap from Stable asset, and complete the swap logic.
-				Self::do_group_and_swap(&mut swaps, SwapLeg::FromStable)?;
+			let mut block_to_process = FirstBlockWithPendingSwaps::<T>::get();
 
-				for swap in swaps {
-					if let Some(swap_output) = swap.final_output {
-						Self::deposit_event(Event::<T>::SwapExecuted {
-							swap_id: swap.swap_id,
-							source_asset: swap.from,
-							destination_asset: swap.to,
-							deposit_amount: swap.amount,
-							swap_input: swap.amount,
-							egress_amount: swap_output,
-							swap_output,
-							intermediate_amount: swap.intermediate_amount(),
-						});
-						// Handle swap completion logic.
-						match &swap.swap_type {
-							SwapType::Swap(destination_address) =>
-								match T::EgressHandler::schedule_egress(
-									swap.to,
-									swap_output,
-									destination_address.clone(),
-									None,
-								) {
-									Ok(ScheduledEgressDetails {
-										egress_id,
-										egress_amount,
-										fee_withheld,
-									}) => {
-										Self::deposit_event(Event::<T>::SwapEgressScheduled {
-											swap_id: swap.swap_id,
-											egress_id,
-											asset: swap.to,
-											amount: egress_amount,
-											fee: fee_withheld,
-										});
-									},
-									Err(err) => {
-										Self::deposit_event(Event::<T>::SwapEgressIgnored {
-											swap_id: swap.swap_id,
-											asset: swap.to,
-											amount: swap_output,
-											reason: err.into(),
-										});
-									},
-								},
-							SwapType::CcmPrincipal(ccm_id) => {
-								Self::handle_ccm_swap_result(
-									*ccm_id,
-									swap_output,
-									CcmSwapLeg::Principal,
-								);
-							},
-							SwapType::CcmGas(ccm_id) => {
-								Self::handle_ccm_swap_result(*ccm_id, swap_output, CcmSwapLeg::Gas);
-							},
-						};
-					} else {
-						debug_assert!(false, "Swap is not completed yet!");
-					}
-				}
-				Ok(())
-			}) {
-				match failed_swap {
-					BatchExecutionError::SwapLegFailed { asset, direction, amount } =>
-						Self::deposit_event(Event::<T>::BatchSwapFailed {
-							asset,
-							direction,
-							amount,
-						}),
-					BatchExecutionError::DispatchError { error } => {
-						log::error!("Failed to execute swap batch: {:?}", error);
-					},
-				}
+			// NOTE: we iterate manually because BlockNumberFor<T> does not implement Step:
+			while block_to_process <= current_block {
+				Self::process_swaps_for_block(block_to_process);
```

Comment:
We want to exit the loop here if a block of swaps fails to execute.
Also we need to rememember to update `FirstBlockWithPendingSwaps`.

We can wrap `process_swaps_for_block` in `#[transactional]` to reduce boilerplate, and do something like this:

```rust
while block_to_process <= current_block {
	match Self::process_swaps_for_block(block_to_process) {
		Err(e) {
			// log error / deposit event
		} else {
			block_to_process += 1
		}
	}
}
FirstBlockWithPendingSwaps::put(block_to_process)
```